### PR TITLE
elf: add uprobe support, closes #127

### LINKED ISSUE #127: elf: add uprobes support
Add support for loading and attaching uprobes / uretprobes.

### PR DESCRIPTION
**Pull request by bhiggins**
Add support for installing uprobe and uretprobe BPF scripts (which is
handled just like kprobes), and for attaching to any number of uprobe
targets.

### CODE REVIEW

File: elf/module.go [hunk 1]
```diff
@@ -205,6 +215,34 @@ func writeKprobeEvent(probeType, eventName, funcName, maxactiveStr string) (int,
...
+	cmd := fmt.Sprintf("%s:%s %s:%#x\n", probeType, eventName, path, offset)
+
+	if _, err = f.WriteString(cmd); err != nil {
+		return -1, fmt.Errorf("cannot write %q to uprobe_events: %v", cmd, err)
+	}
+
+	uprobeIdFile := fmt.Sprintf("/sys/kernel/debug/tracing/events/uprobes/%s/id", eventName)
+	uprobeIdBytes, err := ioutil.ReadFile(uprobeIdFile)
+	if err != nil {
+		return -1, fmt.Errorf("cannot read kprobe id: %v", err)
```
Comment (graphaelli) [inline elf/module.go]: uprobe

File: elf/module.go [hunk 2]
```diff
@@ -462,6 +537,28 @@ func disableKprobe(eventName string) error {
...
+func disableUprobe(eventName string) error {
+	uprobeEventsFileName := "/sys/kernel/debug/tracing/uprobe_events"
+	f, err := os.OpenFile(uprobeEventsFileName, os.O_APPEND|os.O_WRONLY, 0)
+	if err != nil {
+		return fmt.Errorf("cannot open uprobe_events: %v", err)
+	}
+	defer f.Close()
+	cmd := fmt.Sprintf("-:%s\n", eventName)
+	if _, err = f.WriteString(cmd); err != nil {
+		return fmt.Errorf("cannot write %q to kprobe_events: %v", cmd, err)
```
Comment (graphaelli) [inline elf/module.go]: uprobe_events

**Comment by graphaelli** [comment]
@bhiggins Any thoughts given to supporting `FETCHARGS`?  Also, this did not resolve #126 for me, any insights?

**Comment by bhiggins** [comment]
Thanks for the review!

I don't have a deep background with uprobes, so I'm curious if/how FETCHARGS are useful when used with BPF. In other words, I'm wondering if FETCHARGS is more useful when uprobes are used on their own, but can't say for sure. What do you think?

I don't have any insight with #126, unfortunately, though I'm curious what the code looked like. Do you have source for a minimal repro? Do you just see this with uprobes or with kprobes or others?

**Comment by graphaelli** [comment]
> I'm wondering if FETCHARGS is more useful when uprobes are used on their own

Probably, I don't think it's a huge issue to leave this out for now and revisit when there is a need.

> Do you just see this with uprobes or with kprobes or others?

I have not tried with others.  I'll get a minimal repro into #126 soon and continue the discussion there.

--

You might want to update the [readme](https://github.com/iovisor/gobpf/pull/124/files#diff-04c6e90faac2675aa89e2176d2eec7d8R37) and [tests](https://github.com/iovisor/gobpf/pull/124/files#diff-2ec25a06a1049e7c924954681bb2b96d) but I'd be happy to do that as a follow on.  I hope this gets merged soon.

**Comment by bhiggins** [comment]
Thanks for pointing out the lack of docs and tests, now included. I should have done that originally.

**Comment by graphaelli** [comment]
@bhiggins mind rebasing?

**Comment by calavera** [comment]
@bhiggins @graphaelli do you mind if I open a new PR with these changes rebased with master? I can keep the commits here and add an extra commit to the rebasing changes. It would be very interesting to have this in master.

**Comment by graphaelli** [comment]
That would be great, thank you.

**Comment by calavera** [comment]
This PR can be closed now, the code is rebased in #151.

**Comment by schu** [comment]
Merged via #151

**Pull request closed without merging**